src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterCodeAction.cs

```diff
@@ -0,0 +1,101 @@
+            {
+                _semanticDocument = document;
+                _service = service;
+                _expression = expression;
+                _allOccurrences = allOccurrences;
+                _trampoline = trampoline;
+                Title = CreateDisplayText(expression);
+            }
+
+            public override string Title { get; }
+
+            protected override async Task<Document> GetChangedDocumentAsync(CancellationToken cancellationToken)
+            {
+                var changedDocument = await GetChangedDocumentCoreAsync(cancellationToken).ConfigureAwait(false);
+                return await Simplifier.ReduceAsync(changedDocument, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<Document> GetChangedDocumentCoreAsync(CancellationToken cancellationToken)
+            {
+                var solution = await _service.IntroduceParameterAsync(_semanticDocument, _expression, _allOccurrences, _trampoline, cancellationToken).ConfigureAwait(false);
+                return solution.GetDocument(_semanticDocument.Document.Id);
+            }
+
+            private string CreateDisplayText(TExpressionSyntax expression)
+            {
+                var singleLineExpression = _semanticDocument.Document.GetLanguageService<ISyntaxFactsService>().ConvertToSingleLine(expression);
+                var nodeString = singleLineExpression.ToString();
+
+                return CreateDisplayText(nodeString);
+            }
+
+            private string CreateDisplayText(string nodeString)
+            {
+                string formatString;
+                if (_allOccurrences)
+                {
+                    if (_trampoline)
+                    {
+                        formatString = FeaturesResources.Introduce_parameter_and_trampoline_for_all_occurrences_of_0;
+                    }
+                    else
+                    {
+                        formatString = FeaturesResources.Introduce_parameter_for_all_occurrences_of_0;
+                    }
+                }
+                else
+                {
+                    if (_trampoline)
+                    {
+                        formatString = FeaturesResources.Introduce_parameter_and_trampoline_for_0;
+                    }
+                    else
+                    {
+                        formatString = FeaturesResources.Introduce_parameter_for_0;
+                    }
+                }
+                return string.Format(formatString, nodeString);
+            }
+
+            protected ITypeSymbol GetExpressionType(
+                CancellationToken cancellationToken)
+            {
+                var semanticModel = _semanticDocument.SemanticModel;
+                var typeInfo = semanticModel.GetTypeInfo(_expression, cancellationToken);
+
+                return typeInfo.Type ?? typeInfo.ConvertedType ?? semanticModel.Compilation.GetSpecialType(SpecialType.System_Object);
```

Comment:
why? what's going on?

Reply:
While we do have helpers for SymbolInfo that offers to get any symbol: https://github.com/dotnet/roslyn/blob/efee485eda1c504c4f51d73f81ce0caa574a25be/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/SymbolInfoExtensions.cs#L32

That is because we often want to be fault tolerant of ambiguous cases. I am fine with returning object as the type if we cannot get it from type info but I do not think we want to return the converted type? Is there a test case that prompted this?

Reply:
Leftover from when I took from introduce local, should have been removed.